### ISSUE TITLE
Segfault in Torch profiler when CUDA Graph Conditional Nodes are used

### ISSUE THREAD

**Issue by leijurv** [pytorch/pytorch#134308]
### 🐛 Describe the bug

This mostly follows NVIDIA's guide for conditional nodes from [here](https://developer.nvidia.com/blog/dynamic-control-flow-in-cuda-graphs-with-conditional-nodes/). It does a single loop running a single kernel that simply ends the loop. It never segfaults outside the Torch profiler. Additionally, the GDB backtrace (below) shows that the exception is within `libcupti.so` which is profiling-related.

Steps to reproduce: 
1. Write the files below (`setup.py`, `repro.py`, `repro.cu`)
2. `pip install -e .`
3. `python repro.py`

<details>
<summary>Example log</summary>

```
$ python repro.py
Warming up
Profiling
STAGE:2024-08-23 04:41:20 142638:142638 ActivityProfilerController.cpp:314] Completed Stage: Warm Up
Loop inside profiler 0
Loop inside profiler 1
Loop inside profiler 2
Loop inside profiler 3
Loop inside profiler 4
Loop inside profiler 5
Loop inside profiler 6
Loop inside profiler 7
Loop inside profiler 8
Loop inside profiler 9
Loop inside profiler 10
Loop inside profiler 11
Segmentation fault (core dumped)
```

</details>

<details>
<summary>GDB backtrace</summary>

```
Warming up
Profiling
STAGE:2024-08-23 04:42:35 142909:142909 ActivityProfilerController.cpp:314] Completed Stage: Warm Up
Loop inside profiler 0
Loop inside profiler 1
Loop inside profiler 2
Loop inside profiler 3
Loop inside profiler 4
Loop inside profiler 5
Loop inside profiler 6
Loop inside profiler 7
Loop inside profiler 8
Loop inside profiler 9
Loop inside profiler 10
Loop inside profiler 11
Loop inside profiler 12
Loop inside profiler 13
Loop inside profiler 14
Loop inside profiler 15
Loop inside profiler 16
Loop inside profiler 17
Loop inside profiler 18
Loop inside profiler 19
Loop inside profiler 20
Loop inside profiler 21
Loop inside profiler 22
Loop inside profiler 23
Loop inside profiler 24
Loop inside profiler 25

Thread 1 "pt_main_thread" received signal SIGSEGV, Segmentation fault.
0x00007fffb1c3188e in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
(gdb) bt
#0  0x00007fffb1c3188e in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#1  0x00007fff4e513ee9 in ?? () from /usr/local/cuda/targets/x86_64-linux/lib/libcupti.so.12
#2  0x00007fff4e50d3a8 in ?? () from /usr/local/cuda/targets/x86_64-linux/lib/libcupti.so.12
#3  0x00007fffb1a03ead in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#4  0x00007fffb19233b4 in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#5  0x00007fffb1abfbcf in ?? () from /lib/x86_64-linux-gnu/libcuda.so.1
#6  0x00007fff50221825 in ?? () from /usr/local/cuda/targets/x86_64-linux/lib/libcudart.so.12
#7  0x00007fff5026b48e in cudaGraphInstantiate () from /usr/local/cuda/targets/x86_64-linux/lib/libcudart.so.12
#8  0x00007fffba9abc17 in cudaGraphInstantiate(CUgraphExec_st**, CUgraph_st*, CUgraphNode_st**, char*, unsigned long) ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#9  0x00007fffba9ac12c in cudaGraphDemo() ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#10 0x00007fffba9d85ed in void pybind11::detail::argument_loader<>::call_impl<void, void (*&)(), , pybind11::detail::void_type>(void (*&)(), std::integer_sequence<unsigned long>, pybind11::detail::void_type&&) && ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#11 0x00007fffba9d4644 in std::enable_if<std::is_void<void>::value, pybind11::detail::void_type>::type pybind11::detail::argument_loader<>::call<void, pybind11::detail::void_type, void (*&)()>(void (*&)()) && ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#12 0x00007fffba9cf970 in pybind11::cpp_function::initialize<void (*&)(), void, , pybind11::name, pybind11::scope, pybind11::sibling>(void (*&)(), void (*)(), pybind11::name const&, pybind11::scope const&, pybind11::sibling const&)::{lambda(pybind11::detail::function_call&)#3}::operator()(pybind11::detail::function_call&) const ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#13 0x00007fffba9cf9d8 in pybind11::cpp_function::initialize<void (*&)(), void, , pybind11::name, pybind11::scope, pybind11::sibling>(void (*&)(), void (*)(), pybind11::name const&, pybind11::scope const&, pybind11::sibling const&)::{lambda(pybind11::detail::function_call&)#3}::_FUN(pybind11::detail::function_call&) ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#14 0x00007fffba9bcf45 in pybind11::cpp_function::dispatcher(_object*, _object*, _object*) ()
   from /home/ubuntu/cudagraphissue/repro_kernels.cpython-310-x86_64-linux-gnu.so
#15 0x00005555556aedae in ?? ()
#16 0x00005555556a552b in _PyObject_MakeTpCall ()
#17 0x0000555555758b60 in ?? ()
#18 0x000055555569f1c7 in _PyEval_EvalFrameDefault ()
#19 0x0000555555694096 in ?? ()
#20 0x0000555555789f66 in PyEval_EvalCode ()
#21 0x00005555557b4e98 in ?? ()
#22 0x00005555557ae79b in ?? ()
#23 0x00005555557b4be5 in ?? ()
#24 0x00005555557b40c8 in _PyRun_SimpleFileObject ()
#25 0x00005555557b3d13 in _PyRun_AnyFileObject ()
#26 0x00005555557a670e in Py_RunMain ()
#27 0x000055555577cdfd in Py_BytesMain ()
#28 0x00007ffff7c29d90 in __libc_start_call_main (main=main@entry=0x55555577cdc0, argc=argc@entry=2, 
    argv=argv@entry=0x7fffffffe438) at ../sysdeps/nptl/libc_start_call_main.h:58
#29 0x00007ffff7c29e40 in __libc_start_main_impl (main=0x55555577cdc0, argc=2, argv=0x7fffffffe438, 
    init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffe428)
    at ../csu/libc-start.c:392
#30 0x000055555577ccf5 in _start ()
(gdb) 
```
</details>

<details>
<summary>setup.py</summary>

```py
from torch.utils.cpp_extension import BuildExtension, CUDAExtension
from setuptools import setup
import pybind11

setup(
    name="repro",
    ext_modules=[CUDAExtension(
        "repro_kernels",
        ["repro.cu"],
        include_dirs=[pybind11.get_include()],
    )],
    cmdclass={"build_ext": BuildExtension},
)
```

</details>

<details>
<summary>repro.py</summary>

```py
from torch.profiler import profile, record_function, ProfilerActivity
import repro_kernels

print("Warming up")
for i in range(1000):
    repro_kernels.cudaGraphDemo()

print("Profiling")
with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA], with_stack=True) as prof:
    for i in range(100):
        print("Loop inside profiler", i)
        repro_kernels.cudaGraphDemo()
```

</details>

<details>
<summary>repro.cu</summary>

```c++
#include <torch/extension.h>
#include <ATen/ATen.h>
#include <ATen/cuda/CUDAContext.h>

__global__ void loopKernel(cudaGraphConditionalHandle handle) {
	// simulate work
	__nanosleep(1000000);
	// end the loop
	cudaGraphSetConditional(handle, 0);
}

void cudaGraphDemo() {
	auto stream = at::cuda::getCurrentCUDAStream().stream();
	cudaGraph_t graph;
	AT_CUDA_CHECK(cudaGraphCreate(&graph, 0));
	cudaGraphConditionalHandle loopHandle;
	AT_CUDA_CHECK(cudaGraphConditionalHandleCreate(&loopHandle, graph, true, cudaGraphCondAssignDefault));
	cudaGraphNodeParams whileNodeParams = { cudaGraphNodeTypeConditional };
	whileNodeParams.conditional.handle = loopHandle;
	whileNodeParams.conditional.type = cudaGraphCondTypeWhile;
	whileNodeParams.conditional.size = 1;
	cudaGraphNode_t loopNode;
	AT_CUDA_CHECK(cudaGraphAddNode(&loopNode, graph, nullptr, 0, &whileNodeParams));
	cudaStream_t captureStream;
	AT_CUDA_CHECK(cudaStreamCreate(&captureStream));
	AT_CUDA_CHECK(cudaStreamBeginCaptureToGraph(captureStream, whileNodeParams.conditional.phGraph_out[0], nullptr, nullptr, 0, cudaStreamCaptureModeThreadLocal));
	loopKernel<<<1, 1, 0, captureStream>>>(loopHandle);
	AT_CUDA_CHECK(cudaPeekAtLastError());
	AT_CUDA_CHECK(cudaStreamEndCapture(captureStream, nullptr));
	AT_CUDA_CHECK(cudaStreamDestroy(captureStream));
	cudaGraphExec_t instance;
	AT_CUDA_CHECK(cudaGraphInstantiate(&instance, graph, NULL, NULL, 0));
	AT_CUDA_CHECK(cudaGraphLaunch(instance, stream));
	AT_CUDA_CHECK(cudaStreamSynchronize(stream));
	AT_CUDA_CHECK(cudaGraphExecDestroy(instance));
	AT_CUDA_CHECK(cudaGraphDestroy(graph));
}

PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
	m.def("cudaGraphDemo", cudaGraphDemo);
}
```

</details>

### Versions

Note that I installed torch using `pip3 install torch --index-url https://download.pytorch.org/whl/cu124`, otherwise this code fails to run at all due to a linker error due to CUDA Graph Conditional Nodes requiring Cuda 12.4.

<details>
<summary>Versions</summary>

```
Collecting environment information...
PyTorch version: 2.3.1
Is debug build: False
CUDA used to build PyTorch: 12.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jul 29 2024, 16:56:48) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-40-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 PCIe
Nvidia driver version: 550.90.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               26
On-line CPU(s) list:                  0-25
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8480+
CPU family:                           6
Model:                                143
Thread(s) per core:                   1
Core(s) per socket:                   1
Socket(s):                            26
Stepping:                             8
BogoMIPS:                             4000.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd arat vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b fsrm md_clear serialize tsxldtrk avx512_fp16 arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            832 KiB (26 instances)
L1i cache:                            832 KiB (26 instances)
L2 cache:                             104 MiB (26 instances)
L3 cache:                             416 MiB (26 instances)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-25
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Unknown: No mitigations
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] flake8==4.0.1
[pip3] numpy==1.25.2
[pip3] torch==2.3.1
[pip3] torchvision==0.18.1
[pip3] triton==3.0.0
[conda] Could not collect
```
</summary>

cc @ptrblck @msaroufim @robieta @chaekit @aaronenyeshi @guotuofeng @guyang3532 @dzhulgakov @davidberard98 @briancoutinho @sraikund16 @sanrise @mcarilli @ezyang @eellison @penguinwu

**Comment by aaronenyeshi** [comment]
cc @sraikund16

**Comment by leijurv** [comment]
Note that it doesn't segfault every time. I commented out the warmup outside the profiler, but otherwise left it unchanged, and ran it 800 times, there were 261 segfaults (so about a third of the time).

**Comment by leijurv** [comment]
Another data point: It segfaults after the profiler too. The GDB backtrace is unchanged. It's as if the profiler permanenlty corrupts some internal state that causes conditional nodes to sometimes fail, even after the profiling is done.

```py
from torch.profiler import profile, record_function, ProfilerActivity
import repro_kernels

print("Warming up")
for i in range(1000):
    repro_kernels.cudaGraphDemo()

print("Profiling")
with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA], with_stack=True) as prof:
    print("Profiled something")

print("Outside the profiler")
for i in range(1000):
    repro_kernels.cudaGraphDemo()
```

I've run this modified version here ^ 100 times and it segfaulted 54 of those times (all of which were after the profiler). So this actually fails more reliably than my original `repro.py`.

**Comment by cora-codes** [comment]
@sraikund16 do you have a workaround here? Currently, we cannot profile our code which makes it difficult to continue with PyTorch :)

**Comment by aaronenyeshi** [comment]
This seems like a bug between CUDA Graphs and CUPTI, and should be reported to NVIDIA.

**Comment by leijurv** [comment]
Reported here: https://forums.developer.nvidia.com/t/usage-cupti-appears-to-rarely-cause-cuda-graph-conditional-nodes-to-segfault-upon-instantiation/305571

**Comment by leijurv** [comment]
Nvidia says:

```
[veraj](https://forums.developer.nvidia.com/u/veraj)
Moderator
Hi, [@leif5]

We can reproduce your issue internally and the fix will be shipped in an upcoming CUDA release.
Thanks for reporting this !
```

**Comment by sraikund16** [comment]
Closing as is NV issue
